File: src/UpDoc/wwwroot/App_Plugins/UpDoc/src/import-facts.ts
import type { MappingDestination } from './workflow.types.js';

/**
 * Writing import facts into a document being created.
 *
 * An import fact is a value describing the import itself — which file the editor picked —
 * rather than content extracted from that file. It needs its own apply path because the
 * ordinary one resolves a source key against extracted sections, and an import fact is
 * not in there.
 *
 * Shared by both bridge files (entity action and collection action) so the two cannot
 * drift apart. See #41 for extracting the rest of the duplicated apply logic.
 */

/** One entry in a MediaPicker3 value. */
interface MediaPickerEntry {
	key: string;
	mediaKey: string;
	mediaTypeAlias: string;
	crops: unknown[];
	focalPoint: unknown | null;
}

/**
 * Builds a MediaPicker3 value for a single media item.
 *
 * `key` identifies this entry within the picker and is distinct from `mediaKey`, which
 * identifies the media item. Umbraco expects a fresh key per entry.
 *
 * `mediaTypeAlias` is deliberately left empty. Umbraco's MediaPicker3PropertyEditor
 * overwrites it from the real media item on save (UpdateMediaTypeAliases), and explicitly
 * tolerates it being absent on create, so sending a guess would be pointless at best and
 * wrong at worst.
 */
export function buildMediaPickerValue(mediaKey: string): MediaPickerEntry[] {
	return [
		{
			key: crypto.randomUUID(),
			mediaKey,
			mediaTypeAlias: '',
			crops: [],
			focalPoint: null,
		},
	];
}

/**
 * Writes the picked media into a destination field.
 *
 * Only top-level fields are supported: a media picker inside a block would need the
 * block's own value JSON rewritten, which the ordinary apply path handles for text but
 * has no equivalent for here. Block destinations are skipped rather than written wrongly.
 */
export function applyImportFactMedia(
	values: Array<{ alias: string; value: unknown }>,
	dest: MappingDestination,
	mediaKey: string,
): void {
	if (dest.blockKey) {
		console.warn(
			`UpDoc: import-fact mapping to a block property ("${dest.target}") is not supported yet — skipped.`,
		);
		return;
	}

	const value = buildMediaPickerValue(mediaKey);
	const existing = values.find((v) => v.alias === dest.target);

	if (existing) {
		existing.value = value;
	} else {
		values.push({ alias: dest.target, value });
	}
}

File: src/UpDoc/Models/ImportFacts.cs
namespace UpDoc.Models;

/// <summary>
/// Mapping sources that describe the import itself rather than content extracted
/// from it — the file the editor picked, and in future the URL they typed.
///
/// These are not sections, so anything that resolves or validates a mapping source
/// against extracted content has to skip them. The "$" prefix marks them as such.
///
/// Mirrors IMPORT_FACT_SOURCE_FILE in workflow.types.ts. The two must agree: the
/// TypeScript side writes these into map.json, this side reads them back.
/// </summary>
public static class ImportFacts
{
    /// <summary>Marks a map.json source as an import fact rather than a section id.</summary>
    public const string Prefix = "$";

    /// <summary>The file picked for this import.</summary>
    public const string SourceFile = "$sourceFile";

    /// <summary>
    /// Whether a map.json source describes the import rather than naming a section.
    /// </summary>
    public static bool IsImportFact(string? source) =>
        source?.StartsWith(Prefix, StringComparison.Ordinal) == true;
}
